src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -152,6 +225,50 @@ void ZooKeeper::init(ZooKeeperArgs args_)
     }
 }
 
+void ZooKeeper::tryConnectSameAZKeeper()
+{
+    const std::string local_az = args.availability_zone;
+    auto & az_helper = ZooKeeperAvailabilityZoneMap::instance();
+    const auto & hosts = az_helper.shuffleHosts(local_az, args.hosts);
+
+    std::set<std::string> attempted_hosts;
+    for (const auto & host : hosts)
+    {
+        std::string connected_host_az = ZooKeeperAvailabilityZoneMap::AZ_UNKNWON;
+        LOG_INFO(log, "Connecting to ZooKeeper host {}, number of attempted hosts {}", host, attempted_hosts.size());
+
+        attempted_hosts.insert(host);
+        try
+        {
+            Coordination::ZooKeeper::Node node;
+            String host_string = host;
+            node.secure = startsWith(host_string, "secure://");
+            if (node.secure)
+                host_string.erase(0, strlen("secure://"));
+            node.address = Poco::Net::SocketAddress(host_string);
+            impl = std::make_unique<Coordination::ZooKeeper>(Coordination::ZooKeeper::Nodes{node}, args, zk_log);
+            connected_host_az = impl->getAvailabilityZone();
+            az_helper.update(host, connected_host_az);
+            if (local_az == connected_host_az)
+            {
+                LOG_INFO(log, "Successfully connected to a ZooKeeper host {} in the same az {}", host, local_az);
+                return;
+            }
+            // Non optimal case that we need to set a session deadline.
+            auto session_timeout_seconds = impl->setClientSessionDeadline(args.fallback_session_lifetime.min_sec, args.fallback_session_lifetime.max_sec);
+            bool need_try_other_host = az_helper.needTryOtherHost(local_az, attempted_hosts);
```

Comment:
That place is interesting.
As I read this I see that at first run, when client does not know servers zones, the client has to connect to all the servers, because there are some unvisited hosts with unknow az in the list. That could stop early if server with local has been found.
But if we found such local server fast, when it is gone, we will connection to the all rest servers.
Thinks are better when we already know all the servers zones.

Reply:
But I do not understand yet what pice of code is working when client reconnects.
Is `ZooKeeper::init` called for it?

Reply:
>  at first run, when client does not know servers zones, the client has to connect to all the servers

Yes initially client do not know server zones. but see line 252, if we are lucky, on the first attempt, client already connects to a local az keeper host, it stops, and not call `setClientSessionDeadline`.


>  if we found such local server fast, when it is gone, we will connection to the all rest servers

Yes. When local az keeper host is gone, we will try other keepers, in a sorted order (local az is first tried, unknown second, and different az last). as long as one connected, keeper client still get initiazlied.

https://github.com/ClickHouse/ClickHouse/pull/56563/files/2efe7fc4bddc9a7455d97d0b113aea52af881249#diff-17249c3158c8311e98425540c1d48298b695b41fadb9081cafc203a3a0edd0f9R94

> But I do not understand yet what pice of code is working when client reconnects.
Is ZooKeeper::init called for it?

Yes. ZooKeeper::startNewSession() -> Constructor -> and then ZooKeeper::init is always invoked. 

https://github.com/ClickHouse/ClickHouse/blob/2b7346d755cf003bf3ba87bac7df275c568080d4/src/Interpreters/Context.cpp#L2920

Reply:
> But if we found such local server fast, when it is gone, we will connection to the all rest servers.

Yeah we will try to connect to other hosts whose AZ are unknown yet. But once we know their AZ, `needTryOtherHost` will give different answer. do not bother to retry because we know the rest unattempted hosts are in different AZ anyway.